kafka/src/main/java/com/metamx/tranquility/kafka/writer/SynchronizedEventWriter.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.metamx.tranquility.kafka.writer;
+
+import com.metamx.common.logger.Logger;
+
+/**
+ * An EventWriter that tracks the number of events sent to it and provides a waitForFlush() method that blocks until
+ * all of the events have flushed (the callback for each event has been called). This is so that we can guarantee all
+ * events have been pushed to Druid before committing the Kafka offsets.
+ */
+public class SynchronizedEventWriter implements EventWriter
+{
+  private static final Logger log = new Logger(SynchronizedEventWriter.class);
+
+  private final EventWriter writer;
+
+  // used to wait for completion callback from all events after flushing before committing the offset to ZK
+  private final Object flushAwaitLock = new Object();
+  private int pendingEvents = 0; // Use under flushAwaitLock
+
+  public SynchronizedEventWriter(EventWriter writer)
+  {
+    this.writer = writer;
+    writer.init(
+        new EventWriter.EventWriterListener()
+        {
+          @Override
+          public void onSuccess(Object o)
+          {
+            decrementPending();
+          }
+
+          @Override
+          public void onFailure(Throwable throwable, Object o)
+          {
+            decrementPending();
+            log.error(throwable, "Failed to send event: %s", (o == null ? "null" : o.toString()));
```

Comment:
It's usually better to make stuff like this a debug and increment a counter that we report periodically, to avoid log spam if there are a flurry of bad messages. Or report at most one message every minute (or whatever) and also report a counter periodically. Generally anything that avoids potentially huge amounts of log messages.